### PR TITLE
Add support for arbitrary types in function parameters

### LINKED ISSUE #161: Functions that take non-dtype arguments
Make it possible to define sth. like
```rs
fn append(s1: str, s2: str) -> str = "{s1}{s2}"

fn and(b1: bool, b2: bool) -> bool = if b1 then b2 else false
```

It also allows us to use `exchange_rate("USD")` etc. instead of having tens of different functions.

### PR DESCRIPTION
**Pull request by sharkdp**
closes #161

### CODE REVIEW

File: numbat/modules/core/strings.nbt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+use core::scalar
+
+fn str_length(a: str) -> Scalar
+fn str_concat(a: str, b: str) -> str = "{a}{b}"
```
Comment (sharkdp) [inline numbat/modules/core/strings.nbt:4]: Think about better names. Or remove for now.

File: numbat/modules/core/booleans.nbt [hunk 2]
```diff
@@ -0,0 +1,2 @@
+fn and(a: bool, b: bool) = if a then b else false
+fn or(a: bool, b: bool) = if a then true else b
```
Comment (sharkdp) [inline numbat/modules/core/booleans.nbt:2]: Is it worth to add these? Or should we just add logical operators?

Comment (eminence) [reply numbat/modules/core/booleans.nbt]: `and` and `or` are so fundamental to booleans, that I would expect that either these functions or infix operators would be provided out-of-the-box

Comment (sharkdp) [reply numbat/modules/core/booleans.nbt]: I agree: https://github.com/sharkdp/numbat/issues/196

**Pull request merged**